build.assets/tooling/lib/logging/hashicorpLoggerAdapter.go

```diff
@@ -0,0 +1,230 @@
+/*
+Copyright 2022 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package logging
+
+import (
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"strings"
+
+	"github.com/hashicorp/go-hclog"
+	"github.com/sirupsen/logrus"
+)
+
+// All of our other tools use logrus for logging, but the gon package
+// uses https://pkg.go.dev/github.com/hashicorp/go-hclog. This is an
+// adapter that will send hclog output to logrus.
+type HCLogLogrusAdapter struct {
+	name        string
+	logger      *logrus.Logger
+	impliedArgs []interface{}
+}
+
+func NewHCLogLogrusAdapter(logrusLogger *logrus.Logger) HCLogLogrusAdapter {
+	return HCLogLogrusAdapter{
+		logger: logrusLogger,
+	}
+}
+
+// Args are alternating key, val pairs
+// keys must be strings
+// vals can be any type, but display is implementation specific
+// Emit a message and key/value pairs at a provided log level
+func (h HCLogLogrusAdapter) Log(level hclog.Level, msg string, args ...interface{}) {
```

Comment:
Just to make sure: Method receivers are always[1] passed by value. Go will be creating a new copy of  `h` on every method call and operating on that, rather than mutating the actual `h` object. Performance-wise, copying 3 pointers is not a big deal  but I don't _think_ that copying is really the behaviour you want here.

To get around this, use:

```suggestion
func (h *HCLogLogrusAdapter) Log(level hclog.Level, msg string, args ...interface{}) {
```

1: even if sometimes those _values_ are _references_ to other things, like `string`s or slices, etc

Reply:
Thanks, updated. Completely understand the argument here. I originally used pointers (as I typically do on all instanced methods) however golang was throwing a fit about the struct implementing a pointer to the `hclog.Logger` interface rather than the `hclog.Logger` interface. I'm guessing I screwed something up the first time as the compiler is no longer complaining.